SwiftCompilerSources/Sources/Optimizer/FunctionPasses/ConstantCapturePropagation.swift

```diff
@@ -0,0 +1,365 @@
+      md.replace(with: newClosure, context)
+    }
+  }
+  context.erase(instruction: partialApply)
+}
+
+private extension PartialApplyInst {
+
+  /// Returns the callee if this is a `partial_apply` of a thunk which directly forwards all arguments
+  /// to the callee and has no other side-effects.
+  func getCalleeOfForwardingThunkWithDeadCaptures() -> Function? {
+    guard let thunk = referencedFunction,
+          let thunkEntryBlock = thunk.blocks.first
+    else {
+      return nil
+    }
+    let numDeadArguments = self.arguments.count
+    let numAliveArgs = thunkEntryBlock.arguments.count - numDeadArguments
+    let deadCalleeArgs = thunkEntryBlock.arguments.dropFirst(numAliveArgs)
+
+    // TODO: handle non-trivial dead arguments, i.e. accept destroy instructions of such arguments.
+    guard deadCalleeArgs.allSatisfy({ $0.type.isTrivial(in: thunk )}) else {
+      return nil
+    }
+
+    var callee: Function? = nil
+    var returnValue: Value? = nil
+    var errorValue: Value? = nil
+
+    for inst in thunk.instructions {
+      switch inst {
+      case let apply as ApplyInst:
+        guard callee == nil, let c = apply.getCalleeWithForwardedArguments(numArguments: numAliveArgs) else {
+          return nil
+        }
+        callee = c
+        returnValue = apply
+      case let tryApply as TryApplyInst:
+        guard callee == nil,
+              let c = tryApply.getCalleeWithForwardedArguments(numArguments: numAliveArgs)
+        else {
+          return nil
+        }
+        callee = c
+        returnValue = tryApply.normalBlock.arguments.first
+        errorValue = tryApply.errorBlock.arguments.first
+      case let returnInst as ReturnInst:
+        guard let returnValue, returnInst.returnedValue == returnValue else {
+          return nil
+        }
+      case let throwInst as ThrowInst:
+        guard let errorValue, throwInst.thrownValue == errorValue else {
+          return nil
+        }
+      case is TermInst:
+        return nil
+      default:
+        if inst.mayHaveSideEffects {
+          return nil
+        }
+      }
+    }
+    return callee
+  }
+
+  func classifyArgumentsForConstness() -> (constArguments: [Operand], nonConstArguments: [Operand])? {
+    var constArgs = [Operand]()
+    var nonConstArgs = [Operand]()
+    var hasKeypath = false
+    for argOp in argumentOperands {
+      if argOp.value.isConstant(hasKeypath: &hasKeypath) {
+        constArgs.append(argOp)
+      } else {
+        nonConstArgs.append(argOp)
+      }
+    }
+
+    // The optimization is beneficial if we can either get rid of all captures, because this
+    // avoids allocating the context.
+    // Or if we can constant propagate at least one keypath. Keypaths are so expensive that constant
+    // propagating a single keypath is already beneficial.
+    if !constArgs.isEmpty,
+       nonConstArgs.isEmpty || hasKeypath
+    {
+      return (constArgs, nonConstArgs)
+    }
+    return nil
```

Comment:
One additional case that comes to mind is if there is a single non const arg that is an AnyObject as this avoids the allocation as well.

Reply:
> One additional case that comes to mind is if there is a single non const arg that is an AnyObject as this avoids the allocation as well because it is used directly as the context.

@dnadoba could you elaborate on that case if you have a moment? curious to understand what you're suggesting a bit better